server/src/main/java/org/elasticsearch/cluster/coordination/NodeJoinExecutor.java

```diff
@@ -393,16 +394,17 @@ public static void ensureNodesCompatibility(Version joiningNodeVersion, Version
      * to ensure that the minimum transport version of the cluster doesn't go backwards.
      **/
     static void ensureTransportVersionBarrier(
-        TransportVersion joiningTransportVersion,
-        Collection<TransportVersion> existingTransportVersions
+        VersionsWrapper joiningVersionsWrapper,
+        Collection<VersionsWrapper> existingTransportVersions
     ) {
         TransportVersion minClusterTransportVersion = existingTransportVersions.stream()
+            .map(VersionsWrapper::transportVersion)
             .min(Comparator.naturalOrder())
             .orElse(TransportVersion.current());
-        if (joiningTransportVersion.before(minClusterTransportVersion)) {
+        if (joiningVersionsWrapper.transportVersion().before(minClusterTransportVersion)) {
             throw new IllegalStateException(
                 "node with transport version ["
-                    + joiningTransportVersion
+                    + joiningVersionsWrapper.transportVersion()
                     + "] may not join a cluster with minimum transport version ["
                     + minClusterTransportVersion
                     + "]"
```

Comment:
Suspect we should move this logic to `VersionsWrapper` too, although that can happen in a follow-up.

Reply:
I think it'll make sense to move the logic when I add the version barrier for system indices.